xmtp_mls/src/identity/xmtp_id/identity.rs

```diff
@@ -0,0 +1,240 @@
+
+        let builder = SignatureRequestBuilder::new(inbox_id);
+
+        if nonce != 0 || Self::is_existing_inbox(inbox_id, api_client).await? {
+            return Err(IdentityError::LegacyKeyReuse);
+        }
+
+        let mut signature_request = builder
+            .create_inbox(member_identifier, nonce)
+            .add_association(installation_public_key.to_vec().into(), member_identifier)
+            .build();
+
+        // We can pre-sign the request with an installation key signature, since we have access to the key
+        signature_request
+            .add_signature(Box::new(
+                Self::sign_with_installation_key(
+                    signature_request.signature_text(),
+                    signature_keys.private().try_into()?,
+                )
+                .await?,
+            ))
+            .await?;
+
+        signature_request
+            .add_signature(Box::new(
+                Self::sign_with_legacy_key(
+                    signature_request.signature_text(),
+                    legacy_signed_private_key,
+                )
+                .await?,
+            ))
+            .await?;
+
+        let identity_update = signature_request.build_identity_update()?;
+
+        api_client.publish_identity_update(identity_update).await?;
+
+        let identity = Self {
+            inbox_id: inbox_id.clone(),
+            installation_keys: signature_keys,
+            credential: create_credential(inbox_id)?,
+            signature_request: None,
+        };
+
+        Ok(identity)
+    }
+
+    async fn is_existing_inbox<ApiClient: XmtpMlsClient + XmtpIdentityClient>(
+        inbox_id: InboxId,
+        api_client: ApiClientWrapper<ApiClient>,
+    ) -> Result<bool, IdentityError> {
+        let identity_updates = api_client
+            .get_identity_updates_v2(vec![GetIdentityUpdatesV2Filter {
+                inbox_id: inbox_id.to_string(),
+                sequence_id: None,
+            }])
+            .await?;
+
+        let inbox_updates = identity_updates.get(&inbox_id);
+
+        Ok(inbox_updates.is_some())
+    }
+
+    async fn sign_with_installation_key(
+        signature_text: String,
+        installation_private_key: &[u8; 32],
+    ) -> Result<InstallationKeySignature, IdentityError> {
+        let signing_key: SigningKey = SigningKey::from_bytes(installation_private_key);
+        let verifying_key = signing_key.verifying_key();
+        let mut prehashed: Sha512 = Sha512::new();
+        prehashed.update(signature_text.clone());
+        let sig = signing_key
+            .sign_prehashed(prehashed, Some(INSTALLATION_KEY_SIGNATURE_CONTEXT))
+            .unwrap();
+
+        let installation_key_sig = InstallationKeySignature::new(
+            signature_text.clone(),
+            sig.to_vec(),
+            verifying_key.as_bytes().to_vec(),
+        );
+
+        Ok(installation_key_sig)
+    }
+
+    async fn sign_with_legacy_key(
```

Comment:
This probably doesn't need to live on `Identity`. Could just be a plain function